python/pyspark/testing/utils.py

```diff
@@ -209,3 +232,144 @@ def check_error(
+
+        # Create a new list by renaming duplicate
+        # columns by adding prefix '_duplicate_'+index
+        for i in duplicate_col_indices:
+            df_cols[i] = df_cols[i] + "_duplicate_" + str(i)
+
+        # Rename duplicate columns
+        result_df = input_df.toDF(*df_cols)
+
+        return result_df
+
+    def compare_rows(r1: Row, r2: Row):
+        def compare_vals(val1, val2):
+            if isinstance(val1, list) and isinstance(val2, list):
+                return len(val1) == len(val2) and all(
+                    compare_vals(x, y) for x, y in zip(val1, val2)
+                )
+            elif isinstance(val1, Row) and isinstance(val2, Row):
+                return all(compare_vals(x, y) for x, y in zip(val1, val2))
+            elif isinstance(val1, dict) and isinstance(val2, dict):
+                return (
+                    len(val1.keys()) == len(val2.keys())
+                    and val1.keys() == val2.keys()
+                    and all(compare_vals(val1[k], val2[k]) for k in val1.keys())
+                )
+            elif isinstance(val1, float) and isinstance(val2, float):
+                if abs(val1 - val2) > 1e-5:
+                    return False
+            else:
+                if val1 != val2:
+                    return False
+            return True
+
+        if r1 is None and r2 is None:
+            return True
+        elif r1 is None or r2 is None:
+            return False
+
+        return compare_vals(r1, r2)
+
+    def assert_schema_equal(
+        df_schema: StructType,
+        expected_schema: StructType,
+    ):
+        if df_schema != expected_schema:
+            raise PySparkAssertionError(
+                error_class="DIFFERENT_SCHEMA",
+                message_parameters={"df_schema": df_schema, "expected_schema": expected_schema},
+            )
+
+    def assert_rows_equal(rows1: Row, rows2: Row):
+        zipped = list(zip_longest(rows1, rows2))
+        rows_equal = True
+        error_msg = "Results do not match: "
+        diff_msg = ""
+        diff_rows_cnt = 0
+
+        for r1, r2 in zipped:
+            if not compare_rows(r1, r2):
+                rows_equal = False
+                diff_rows_cnt += 1
+                diff_msg += (
+                    "[df]" + "\n" + str(r1) + "\n\n" + "[expected]" + "\n" + str(r2) + "\n\n"
+                )
+                diff_msg += "********************" + "\n\n"
+
+        if not rows_equal:
+            percent_diff = diff_rows_cnt / len(zipped)
+            error_msg += "( %.5f %% )" % percent_diff
+            error_msg += "\n" + diff_msg
+            raise PySparkAssertionError(
+                error_class="DIFFERENT_ROWS",
+                message_parameters={"error_msg": error_msg},
+            )
+
+    if ignore_row_order:
+        try:
+            df = rename_duplicate_cols(df)
+            expected = rename_duplicate_cols(expected)
```

Comment:
Do we need this?
I think `assert_schema_equal` will check the duplicated names properly, and `assert_rows_equal` doesn't need to check column/field names anymore.

Reply:
If we need, the current implementation is not enough because it doesn't handle nested struct field names.

Reply:
it's needed for `df.sort`, but cleaned up the code to retain old df column names